itou/www/autocomplete/views.py

```diff
@@ -38,26 +39,31 @@ def cities_autocomplete(request):
-                # are not expected, for instance results containing letters not present in the search.
-                # It has been decided with the UX to use the simplest approach. Secondly, it seems
-                # that most people look for a city by "the start of the name", not by "any word within
-                # the name" so the icontains lookup, ordered by index of the matching string, feels more
-                # natural.
-                # The hyphenated/unhyphenated thing has been added considering the mess that hyphens
-                # represent in french city names. It is ugly, but simpler than the other approach which
-                # could have been to look for the slugs, but requires a perfect sync of th slug and name
-                # at all times, just for the search.
+            # The double search does not seem ideal, but we *did* start with a trigram similarity
+            # and word similarity approach. It is disappointing since it does return results that
+            # are not expected, for instance results containing letters not present in the search.
+            # It has been decided with the UX to use the simplest approach. Secondly, it seems
+            # that most people look for a city by "the start of the name", not by "any word within
+            # the name" so the icontains lookup, ordered by index of the matching string, feels more
+            # natural.
+            # The hyphenated/unhyphenated thing has been added considering the mess that hyphens
+            # represent in french city names. It is ugly, but simpler than the other approach which
+            # could have been to look for the slugs, but requires a perfect sync of th slug and name
+            # at all times, just for the search.
+            cities = (
                 City.objects.filter(
                     Q(name__unaccent__icontains=term_spaced) | Q(name__unaccent__icontains=term_hyphenated)
-                )[:MAX_CITIES_TO_RETURN],
-                # - the results __starting__ by the searched term are favoured (Paris over Cormeil-en-Parisis)
-                # - then if the length is the same, by alphabetic order
-                # - then if everything is the same (Sainte-Colombe...) by department.
-                key=lambda c: (unidecode(c.name.lower()).index(index_term), c.name, c.department),
+                )
+                .annotate(
+                    spaced_index_with_0=StrIndex(Lower("name__unaccent"), Value(term_spaced)),
+                    hyphenated_index_with_0=StrIndex(Lower("name__unaccent"), Value(term_hyphenated)),
+                    spaced_index=Case(When(spaced_index_with_0=0, then=None), default=F("spaced_index_with_0")),
+                    hyphenated_index=Case(
+                        When(hyphenated_index_with_0=0, then=None), default=F("hyphenated_index_with_0")
+                    ),
```

Comment:
Vu que tu utilises `Least`, est-ce que [`NullIf`](https://docs.djangoproject.com/en/4.1/ref/models/database-functions/#nullif) ne permettrais pas de supprimer le `Case` ?
```suggestion
                    spaced_index=NullIf(StrIndex(Lower("name__unaccent"), Value(term_spaced)), 0),
                    hyphenated_index=NullIf(StrIndex(Lower("name__unaccent"), Value(term_hyphenated)), 0),
```